### PR TITLE
IDEA-352355 support MutableSharedFlow and MutableStateFlow

### PR DESCRIPTION
**Pull request by AlexVanGogen**
Corresponding change in coroutines: https://github.com/JetBrains/intellij-deps-kotlinx.coroutines/pull/4

### CODE REVIEW

**Comment by cypok** [comment]
LGTM

**Review by gorrus — CHANGES_REQUESTED** [review]

File: src/main/java/com/intellij/rt/debugger/agent/SharedFlowTransformer.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+class SharedFlowTransformer implements ClassFileTransformer {
+    @Override
+    public byte[] transform(final ClassLoader loader, String className, Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) {
+        if (!"kotlinx/coroutines/flow/SharedFlowImpl".equals(className)) {
+            return classfileBuffer;
+        }
+        ClassReader reader = new ClassReader(classfileBuffer);
+        ClassWriter writer = new ClassWriter(reader, ClassWriter.COMPUTE_FRAMES) {
+            @Override
+            protected ClassLoader getClassLoader() {
```
Comment (gorrus) [inline src/main/java/com/intellij/rt/debugger/agent/SharedFlowTransformer.java:17]: is it still needed?

**Review by gorrus — CHANGES_REQUESTED** [review]

File: src/main/java/com/intellij/rt/debugger/agent/SharedFlowTransformer.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import java.security.ProtectionDomain;
+
+class SharedFlowTransformer implements ClassFileTransformer {
+    @Override
+    public byte[] transform(final ClassLoader loader, String className, Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) {
+        if (!"kotlinx/coroutines/flow/SharedFlowImpl".equals(className)) {
+            return classfileBuffer;
+        }
+        ClassReader reader = new ClassReader(classfileBuffer);
+        ClassWriter writer = new ClassWriter(reader, ClassWriter.COMPUTE_FRAMES) {
```
Comment (gorrus) [inline src/main/java/com/intellij/rt/debugger/agent/SharedFlowTransformer.java:15]: is ClassWriter.COMPUTE_FRAMES still needed?

File: src/main/java/com/intellij/rt/debugger/agent/SharedFlowTransformer.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                switch (name) {
+                    case "wrap":
+                        return new WrapMethodTransformer(mv);
+                    case "unwrap":
+                        return new UnwrapMethodTransformer(mv);
+                }
+                return mv;
+            }
+        }, 0);
+        return writer.toByteArray();
```
Comment (gorrus) [inline src/main/java/com/intellij/rt/debugger/agent/SharedFlowTransformer.java:34]: maybe call storeClassForDebug for debug purposes

**Review by gorrus — APPROVED** [review]

**Pull request merged**